File: model/subproblem.go
package model

type Subproblem struct {
	id          SubproblemID
	position    int
	description string
	language    string
	optional    bool
	grade       int
}

func NewSubproblem(id SubproblemID, position int, description, language string, optional bool, grade int) *Subproblem {
	return &Subproblem{
		id:          id,
		position:    position,
		description: description,
		language:    language,
		optional:    optional,
		grade:       grade,
	}
}

func (s *Subproblem) ID() SubproblemID {
	return s.id
}

func (s *Subproblem) Position() int {
	return s.position
}

func (s *Subproblem) Description() string {
	return s.description
}

func (s *Subproblem) Language() string {
	return s.language
}

func (s *Subproblem) Optional() bool {
	return s.optional
}

func (s *Subproblem) Grade() int {
	return s.grade
}

File: view/pages/problem_edit/reducers/Subproblem.js
// @flow

import type { Action } from '../actions';
import {
  ADD_SUBPROBLEM,
  EDIT_SUBPROBLEM,
  REMOVE_SUBPROBLEM
} from '../actions';
import type { ReduxState, Subproblem } from '../types';
import { assign, generateID } from '../../../utils';
import { resolveSortedSubproblemList } from '../utils';

export default function subproblemReducer(state: ReduxState, action: Action): ReduxState {
  switch (action.type) {
  case ADD_SUBPROBLEM: {
    const position = Object.getOwnPropertyNames(state.subproblems).length+1;
    const subproblem = newSubproblem(position);
    const subproblems = assign(state.subproblems, { [subproblem.id]: subproblem });
    return assign(state, { subproblems });
  }

  case EDIT_SUBPROBLEM: {
    const { id, prop, value } = action.payload;
    const prevSubproblem = state.subproblems[id];
    if (prevSubproblem == null) {
      return state;
    }

    const subproblem = assign(prevSubproblem, { [prop]: value });
    const subproblems = assign(state.subproblems, { [id]: subproblem });
    return assign(state, { subproblems });
  }

  case REMOVE_SUBPROBLEM: {
    const { id } = action.payload;

    var subproblems = assign(state.subproblems, {});
    delete subproblems[id];

    // Refresh positions.
    const sortedSubproblems = resolveSortedSubproblemList(subproblems);
    var position = 1;
    for (var subproblem of sortedSubproblems) {
      subproblems[subproblem.id].position = position;
      position++;
    }

    return assign(state, { subproblems });
  }

  default:
    return state;
  }
}

function newSubproblem(position: number): Subproblem {
  return {
    id: generateID(),
    position,
    description: '',
    language: 'java',
    grade: 0,
    optional: false,
    testIDs: [],
  };
}

